libpod/image/pull.go

```diff
@@ -149,7 +160,9 @@ func (ir *Runtime) getPullListFromRef(ctx context.Context, srcRef types.ImageRef
 		image := splitArr[1]
 		// remove leading "/"
 		if image[:1] == "/" {
-			image = image[1:]
+			// Instead of removing the leading /, set localhost as the registry
+			// so docker.io isn't prepended, and the path becomes the repository
+			image = DefaultLocalRepo + image
```

Comment:
The line this is commented on is for DirTransports, where before the path to the file was being interpreted as the registry, and then code used to parse docker archives jumped in and threw the docker.io. The one above in {getPullStruct, getPullRefPair} is for parsing an archive with no manifest information (e.g. the tar file @edsantiago provided in #667 ). They're slightly different cases. Good point on the comment I'll fix that.

Reply:
> The one above in {getPullStruct, getPullRefPair} is for parsing an archive with no manifest information 

If the intent was to change _only_ that case, then the change does not belong in `getPullRefPair`, which is called in _all_ the cases handled by `getPullListFromRef`; it should happen only inside the `srcRef.Transport().Name() == DockerArchive` case.

Note in particular how the `getPullRefPair` change impacts the `docker://busybox` input (which is, and must remain, equivalent to `docker://docker.io/library/busybox`, but the pulled image is currently saved as `localhost/busybox`).

---

As for this particular line being redundant with `getPullRefPair`: (All values untested, just from reading the code. I hope I haven’t overlooked anything.)

AFAICS the original version parsed `dir:/foo` to `image == "foo"`, and then the updated `getPullRefPair` turned that into `destName == "localhost/foo"`. The same thing happened with `dir:foo`.

The updated version parses `dir:/foo` to `image == "localhost/foo"`, and then `getPullRefPair` tries to optionally add `localhost` _again_ (that’s the redundant part). Meanwhile, `dir:foo` has `image == "foo"`, and we rely on `getPullRefPair` to add `localhost` to `destName`. 

Admittedly, the update _does_ make a difference for `dir:/foo/bar`, where the original parsed this into `image == "foo/bar"`, and `getPullRefPair` did nothing because the `imageParts.hasRegistry` semantics is surprising, resulting in `destName == "foo/bar"`; the updated version makes `image == destName == "localhost/foo/bar"`. But then consider `dir:foo/bar`, which is not affected by this line change and ends up with `image == destName == "foo/bar"`, both with the original and updated version of this line.

 

Ultimately, though, it’s pretty unclear to me how much of the path should be preserved, and how.  If the path happens to be `/example.com/foo/bar`, should the final name be `example.com/foo/bar` or `localhost/example.com/foo/bar`? (Both are valid docker-like names.) And the same question for `example.com/foo/bar`.

(It’s also a bit weird that `dir:foo` and `dir:$(pwd)/foo` result in wildly different names. Would anyone ever want the image name to be`localhost/home/mitr/work/project/foo`?)

One thing I feel fairly sure about: it does not really make much sense to use a different name conversion for relative and absolute path inputs. So, strip the leading `/` first, and _then_ apply whatever policy is desired to turn the directory path into a docker-like reference. [If it is possible at all; e.g. upper-case letters are forbidden in references! Not that I have a much better idea.]

Overall, the desired state is probably up to @rhatdan.

Reply:
> Note in particular how the `getPullRefPair` change impacts the `docker://busybox` input (which is, and must remain, equivalent to `docker://docker.io/library/busybox`, but the pulled image is currently saved as `localhost/busybox`).

I was mistaken about this; `docker://busybox` is passed as `destName` to `getPullRefPair`, where parsing it as a reference, or course, fails, so `localhost/` is not prepended, and the destination uses the `.DockerReference()` value, which is correct.